presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java

```diff
@@ -288,25 +282,107 @@ public void setWindowFunctions(QuerySpecification node, List<FunctionCall> funct
         return windowFunctions.get(query);
     }
 
-    public void setOutputDescriptor(RelationType descriptor)
+    public Scope getScope(Node node)
     {
-        outputDescriptor = descriptor;
+        return tryGetScope(node).orElseThrow(() -> new IllegalArgumentException(String.format("Analysis does not contain information for node: %s", node)));
     }
 
-    public RelationType getOutputDescriptor()
+    public Optional<Scope> tryGetScope(Node node)
+    {
+        // TODO check performance impact for planning of getScope method
+        if (scopes.containsKey(node)) {
+            return Optional.of(scopes.get(node));
+        }
+
+        Optional<Node> root = roots.stream()
+                .filter(i -> nodeContains(i, node))
+                .findAny();
+
+        if (root.isPresent()) {
+            GetScopeVisitor visitor = new GetScopeVisitor(scopes, node);
+            visitor.process(root.get(), new LinkedList<>());
+            return visitor.getResult();
+        }
+        return Optional.empty();
+    }
+
+    public Node getRoot()
     {
-        return outputDescriptor;
+        return roots.get(0);
     }
 
-    public void setOutputDescriptor(Node node, RelationType descriptor)
+    public Scope getRootScope()
+    {
+        return getScope(getRoot());
+    }
+
+    private static class GetScopeVisitor
```

Comment:
Such cases have to be treated separately. Each of such statement (or procedure) have to get special parent scope which will contain information produced by previous statements (or procedures). Such cases already exists in the code. See `StatementAnalyzer#visitQuerySpecyfication` or `StatementAnalyzer#analyzeWith`. 

QuerySpecyfication can be treated as sequence of following "statements": 
 - scan (FROM clause)
 - filter (WHERE cluase)
 - aggregation 
 - projection
Each of above is manipulating on scope produced by previous statement. In code it is simplified (scan and filter uses one `sourceScope`  as well as aggregation and projection uses one `outputScope`).

Lets take a look how `StatementAnalyzer#analyzeWith` works.
```
- main query
 - with
   - a - query1
   - b - query2
   - c - query3
```

In above example when analyzing `query1`, `query2` and `query3` each of the queries gets different parent scope:
 - `query1` gets original parent scope
 - `query2` gets original parent scope with results from `query1`
 - `query3` gets parent scope of `query2` with added results of `query2`.
Then to `with` node I bound the `scope`  which contains results from all the queries (parent scope of `query3` with added results of `query3`).